File: app/src/main/java/com/example/taxconnect/ExploreCAsActivity.java
package com.example.taxconnect;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.text.Editable;
import android.text.TextWatcher;
import android.widget.ArrayAdapter;
import android.widget.Toast;
import android.widget.EditText;
import androidx.appcompat.app.AlertDialog;
import android.text.InputType;
import android.widget.LinearLayout;
import android.content.DialogInterface;
import android.graphics.Typeface;
import android.text.SpannableString;
import android.text.Spanned;
import android.text.style.StyleSpan;
import com.google.android.material.chip.Chip;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.taxconnect.adapter.CAAdapter;
import com.example.taxconnect.databinding.ActivityExploreCasBinding;
import com.example.taxconnect.model.UserModel;
import com.google.firebase.firestore.DocumentSnapshot;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;
import java.util.LinkedHashSet;
import java.util.Set;

public class ExploreCAsActivity extends AppCompatActivity implements CAAdapter.OnConnectClickListener {

    private ActivityExploreCasBinding binding;
    private DataRepository repository;
    private CAAdapter adapter;
    private List<UserModel> fullCAList = new ArrayList<>();
    private double filterMinPrice = 0;
    private double filterMaxPrice = Double.MAX_VALUE;
    private String currentSearchQuery = "";
    private boolean filterVerifiedOnly = false;
    private android.content.SharedPreferences prefs;
    private static final String PREF_RECENT_SEARCHES = "recent_searches";
    private final android.os.Handler searchHandler = new android.os.Handler(android.os.Looper.getMainLooper());
    private final java.lang.Runnable searchRunnable = new java.lang.Runnable() {
        @Override
        public void run() {
            applySearch(currentSearchQuery);
        }
    };
    private java.util.Set<String> favoriteIds = new java.util.HashSet<>();
    private boolean favoritesOnly = false;
    private String currentUserCity = null;
    private DocumentSnapshot lastCaSnapshot = null;
    private boolean isLoadingCAs = false;
    private boolean hasMoreCAs = true;
    private static final int CA_PAGE_SIZE = 30;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityExploreCasBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        repository = DataRepository.getInstance();
        prefs = getSharedPreferences("explore_prefs", MODE_PRIVATE);
        
        setupUI();
        setupRecyclerView();
        setupSorting();
        setupSearch();
        renderRecentChips(loadRecentSearches());
        showLoading(true);
        fetchCAList();
        loadFavorites();
        loadCurrentUserCity();
        
        String initialQuery = getIntent().getStringExtra("QUERY");
        if (initialQuery != null && !initialQuery.isEmpty()) {
            binding.etSearch.setText(initialQuery);
            binding.etSearch.setSelection(initialQuery.length());
            saveRecentSearch(initialQuery);
            renderRecentChips(loadRecentSearches());
        }
        favoritesOnly = getIntent().getBooleanExtra("FAVORITES_ONLY", false);
    }

    private void setupUI() {
        binding.ivBack.setOnClickListener(v -> finish());
        binding.btnFilter.setOnClickListener(v -> showPriceFilterDialog());
        binding.chipVerified.setOnCheckedChangeListener((buttonView, isChecked) -> {
            filterVerifiedOnly = isChecked;
            filterList(currentSearchQuery);
        });
    }

    private void setupRecyclerView() {
        adapter = new CAAdapter(new ArrayList<>(), this, this, false);
        binding.rvCAList.setLayoutManager(new LinearLayoutManager(this));
        binding.rvCAList.setAdapter(adapter);
        binding.rvCAList.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                if (dy <= 0) return;
                LinearLayoutManager layoutManager = (LinearLayoutManager) recyclerView.getLayoutManager();
                if (layoutManager == null) return;
                int lastVisible = layoutManager.findLastVisibleItemPosition();
                int total = adapter.getItemCount();
                if (!isLoadingCAs && hasMoreCAs && lastVisible >= total - 5) {
                    loadMoreCAs(false);
                }
            }
        });
        adapter.setOnSaveClickListener(ca -> {
            String uid = com.google.firebase.auth.FirebaseAuth.getInstance().getUid();
            if (uid == null) return;
            repository.toggleFavorite(uid, ca.getUid(), new DataRepository.DataCallback<Boolean>() {
                @Override
                public void onSuccess(Boolean isFav) {
                    if (isFav) {
                        favoriteIds.add(ca.getUid());
                    } else {
                        favoriteIds.remove(ca.getUid());
                    }
                    adapter.setFavorites(favoriteIds);
                }
                @Override
                public void onError(String error) {}
            });
        });
    }

    private void showPriceFilterDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle("Filter by Price Range");

        LinearLayout layout = new LinearLayout(this);
        layout.setOrientation(LinearLayout.VERTICAL);
        layout.setPadding(32, 16, 32, 16);

        final EditText inputMin = new EditText(this);
        inputMin.setInputType(InputType.TYPE_CLASS_NUMBER | InputType.TYPE_NUMBER_FLAG_DECIMAL);
        inputMin.setHint("Min Price (₹)");
        if (filterMinPrice > 0) inputMin.setText(String.valueOf(filterMinPrice));
        layout.addView(inputMin);

        final EditText inputMax = new EditText(this);
        inputMax.setInputType(InputType.TYPE_CLASS_NUMBER | InputType.TYPE_NUMBER_FLAG_DECIMAL);
        inputMax.setHint("Max Price (₹)");
        if (filterMaxPrice < Double.MAX_VALUE) inputMax.setText(String.valueOf(filterMaxPrice));
        layout.addView(inputMax);

        builder.setView(layout);

        builder.setPositiveButton("Apply", (dialog, which) -> {
            String minStr = inputMin.getText().toString();
            String maxStr = inputMax.getText().toString();

            try {
                filterMinPrice = minStr.isEmpty() ? 0 : Double.parseDouble(minStr);
            } catch (NumberFormatException e) {
                filterMinPrice = 0;
            }

            try {
                filterMaxPrice = maxStr.isEmpty() ? Double.MAX_VALUE : Double.parseDouble(maxStr);
            } catch (NumberFormatException e) {
                filterMaxPrice = Double.MAX_VALUE;
            }

            // Re-apply filters
            filterList(currentSearchQuery);
        });

        builder.setNegativeButton("Clear", (dialog, which) -> {
            filterMinPrice = 0;
            filterMaxPrice = Double.MAX_VALUE;
            filterList(currentSearchQuery);
            dialog.cancel();
        });
        
        builder.setNeutralButton("Cancel", (dialog, which) -> dialog.cancel());

        builder.show();
    }

    private void setupSorting() {
        String[] sortOptions = {"Rating (High to Low)", "Experience (High to Low)", "Charges (Low to High)"};
        ArrayAdapter<String> sortAdapter = new ArrayAdapter<>(this, R.layout.item_spinner, sortOptions);
        sortAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        binding.spinnerSort.setAdapter(sortAdapter);

        binding.spinnerSort.setOnItemSelectedListener(new android.widget.AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(android.widget.AdapterView<?> parent, View view, int position, long id) {
                sortList(position);
            }

            @Override
            public void onNothingSelected(android.widget.AdapterView<?> parent) {}
        });
    }

    private void setupSearch() {
        binding.etSearch.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                currentSearchQuery = s.toString();
                debounceSearch();
            }

            @Override
            public void afterTextChanged(Editable s) {}
        });
    }

    private void fetchCAList() {
        fullCAList.clear();
        lastCaSnapshot = null;
        hasMoreCAs = true;
        loadMoreCAs(true);
    }

    private void loadMoreCAs(boolean initialLoad) {
        if (isLoadingCAs || !hasMoreCAs) return;
        isLoadingCAs = true;
        if (initialLoad) {
            showLoading(true);
        }

        repository.getCaPage(CA_PAGE_SIZE, lastCaSnapshot, new DataRepository.DataCallback<DataRepository.PageResult<UserModel>>() {
            @Override
            public void onSuccess(DataRepository.PageResult<UserModel> data) {
                if (initialLoad) {
                    showLoading(false);
                }
                if (data == null || data.getItems() == null) {
                     binding.tvEmptyState.setVisibility(View.VISIBLE);
                     binding.rvCAList.setVisibility(View.GONE);
                     binding.tvRegisteredCount.setText("0 Registered CAs");
                    isLoadingCAs = false;
                     return;
                }
                
                List<UserModel> newItems = data.getItems();
                if (newItems.isEmpty()) {
                    hasMoreCAs = false;
                } else {
                    fullCAList.addAll(newItems);
                    lastCaSnapshot = data.getLastSnapshot();
                }
                
                String countText = String.valueOf(fullCAList.size());
                String fullText = countText + " Registered CAs";
                SpannableString spannable = new SpannableString(fullText);
                spannable.setSpan(new StyleSpan(Typeface.BOLD), 0, countText.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                binding.tvRegisteredCount.setText(spannable);
                
                if (fullCAList.isEmpty()) {
                    binding.tvEmptyState.setVisibility(View.VISIBLE);
                    binding.rvCAList.setVisibility(View.GONE);
                } else {
                    binding.tvEmptyState.setVisibility(View.GONE);
                    binding.rvCAList.setVisibility(View.VISIBLE);
                    sortList(binding.spinnerSort.getSelectedItemPosition());
                    adapter.notifyDataSetChanged();
                }
                isLoadingCAs = false;
            }

            @Override
            public void onError(String error) {
                if (initialLoad) {
                    showLoading(false);
                }
                Toast.makeText(ExploreCAsActivity.this, "Error fetching CAs: " + error, Toast.LENGTH_SHORT).show();
                binding.tvRegisteredCount.setText("0 Registered CAs");
                binding.tvEmptyState.setVisibility(View.VISIBLE);
                binding.rvCAList.setVisibility(View.GONE);
                isLoadingCAs = false;
            }
        });
    }

    private void filterList(String query) {
        currentSearchQuery = query;
        List<UserModel> filtered = fullCAList.stream()
                .filter(ca -> {
                    boolean matchesQuery = query.isEmpty() ||
                            (ca.getName() != null && ca.getName().toLowerCase().contains(query.toLowerCase())) ||
                            (ca.getSpecialization() != null && ca.getSpecialization().toLowerCase().contains(query.toLowerCase())) ||
                            (ca.getCity() != null && ca.getCity().toLowerCase().contains(query.toLowerCase()));
                    
                    boolean matchesPrice = true;
                    if (ca.getMinCharges() != null) {
                        try {
                            String num = ca.getMinCharges().replaceAll("[^0-9.]", "");
                            if (!num.isEmpty()) {
                                double price = Double.parseDouble(num);
                                matchesPrice = price >= filterMinPrice && price <= filterMaxPrice;
                            }
                        } catch (NumberFormatException e) {
                            // Ignore
                        }
                    }
                    boolean matchesVerified = !filterVerifiedOnly || ca.isVerified();
                    return matchesQuery && matchesPrice && matchesVerified;
                })
                .collect(Collectors.toList());

        if (filtered.isEmpty()) {
             binding.tvEmptyState.setVisibility(View.VISIBLE);
             binding.rvCAList.setVisibility(View.GONE);
        } else {
             binding.tvEmptyState.setVisibility(View.GONE);
             binding.rvCAList.setVisibility(View.VISIBLE);
        }
        adapter.updateList(filtered);
    }
    
    private void sortList(int position) {
        if (fullCAList == null) return;
        
        switch (position) {
            case 0: // Rating (High to Low)
                Collections.sort(fullCAList, (o1, o2) -> Double.compare(o2.getRating(), o1.getRating()));
                break;
            case 1: // Experience (High to Low)
                Collections.sort(fullCAList, (o1, o2) -> {
                    double e1 = parseExperience(o1.getExperience());
                    double e2 = parseExperience(o2.getExperience());
                    return Double.compare(e2, e1);
                });
                break;
            case 2: // Charges (Low to High)
                Collections.sort(fullCAList, (o1, o2) -> {
                    double c1 = parseCharges(o1.getMinCharges());
                    double c2 = parseCharges(o2.getMinCharges());
                    return Double.compare(c1, c2);
                });
                break;
        }
        filterList(currentSearchQuery);
    }

    private double parseExperience(String exp) {
        if (exp == null) return 0;
        try {
            return Double.parseDouble(exp.replaceAll("[^0-9.]", ""));
        } catch (NumberFormatException e) {
            return 0;
        }
    }

    private double parseCharges(String charge) {
        if (charge == null) return Double.MAX_VALUE;
        try {
            return Double.parseDouble(charge.replaceAll("[^0-9.]", ""));
        } catch (NumberFormatException e) {
            return Double.MAX_VALUE;
        }
    }

    private void debounceSearch() {
        searchHandler.removeCallbacks(searchRunnable);
        searchHandler.postDelayed(searchRunnable, 300);
    }

    private void applySearch(String query) {
        filterList(query);
        if (query != null && query.trim().length() >= 2) {
            saveRecentSearch(query.trim());
            renderRecentChips(loadRecentSearches());
        }
    }

    private List<String> loadRecentSearches() {
        String csv = prefs.getString(PREF_RECENT_SEARCHES, "");
        List<String> result = new ArrayList<>();
        if (csv == null || csv.isEmpty()) return result;
        for (String s : csv.split("\\|")) {
            if (!s.trim().isEmpty()) result.add(s.trim());
        }
        return result;
    }

    private void saveRecentSearch(String query) {
        List<String> current = loadRecentSearches();
        Set<String> set = new LinkedHashSet<>(current);
        // Move to front: remove then add
        set.remove(query);
        set.add(query);
        // Keep last 5 (preserve insertion order, we will trim if size > 5)
        List<String> ordered = new ArrayList<>(set);
        if (ordered.size() > 5) {
            ordered = ordered.subList(ordered.size() - 5, ordered.size());
        }
        String csv = String.join("|", ordered);
        prefs.edit().putString(PREF_RECENT_SEARCHES, csv).apply();
    }

    private void clearRecentSearches() {
        prefs.edit().remove(PREF_RECENT_SEARCHES).apply();
        renderRecentChips(new ArrayList<>());
    }

    private void renderRecentChips(List<String> items) {
        binding.chipGroupRecent.removeAllViews();
        binding.hsRecentSearches.setVisibility(items.isEmpty() ? View.GONE : View.VISIBLE);
        for (String label : items) {
            Chip chip = new Chip(this);
            chip.setText(label);
            chip.setCheckable(false);
            chip.setClickable(true);
            chip.setOnClickListener(v -> {
                binding.etSearch.setText(label);
                binding.etSearch.setSelection(label.length());
                applySearch(label);
            });
            binding.chipGroupRecent.addView(chip);
        }
        if (!items.isEmpty()) {
            Chip clearChip = new Chip(this);
            clearChip.setText("Clear");
            clearChip.setCheckable(false);
            clearChip.setClickable(true);
            clearChip.setOnClickListener(v -> clearRecentSearches());
            binding.chipGroupRecent.addView(clearChip);
        }
    }

    private void showLoading(boolean show) {
        binding.progressBar.setVisibility(show ? View.VISIBLE : View.GONE);
        binding.rvCAList.setVisibility(show ? View.GONE : View.VISIBLE);
        binding.tvEmptyState.setVisibility(View.GONE);
    }

    private void loadFavorites() {
        String uid = com.google.firebase.auth.FirebaseAuth.getInstance().getUid();
        if (uid == null) return;
        repository.getFavoriteCaIds(uid, new DataRepository.DataCallback<List<String>>() {
            @Override
            public void onSuccess(List<String> ids) {
                favoriteIds.clear();
                favoriteIds.addAll(ids);
                adapter.setFavorites(favoriteIds);
            }
            @Override
            public void onError(String error) {}
        });
    }

    private void loadCurrentUserCity() {
        String uid = com.google.firebase.auth.FirebaseAuth.getInstance().getUid();
        if (uid == null) return;
        repository.fetchUser(uid, new DataRepository.DataCallback<UserModel>() {
            @Override
            public void onSuccess(UserModel user) {
                if (user != null) {
                    currentUserCity = user.getCity();
                }
            }
            @Override
            public void onError(String error) {}
        });
    }

    @Override
    public void onConnectClick(UserModel ca) {
        Intent intent = new Intent(ExploreCAsActivity.this, CADetailActivity.class);
        intent.putExtra("CA_DATA", ca);
        startActivity(intent);
    }
}

File: tax connect/app/src/main/java/com/example/taxconnect/features/ca/ExploreCAsActivity.kt
package com.example.taxconnect.features.ca

import android.content.Intent
import android.graphics.Typeface
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.text.*
import android.text.style.StyleSpan
import android.view.LayoutInflater
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.*
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.taxconnect.R
import com.example.taxconnect.features.ca.CAAdapter
import com.example.taxconnect.databinding.ActivityExploreCasBinding
import com.example.taxconnect.data.models.UserModel
import com.example.taxconnect.data.repositories.DataRepository
import com.google.android.material.chip.Chip
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.DocumentSnapshot
import java.util.*
import java.util.stream.Collectors
import com.example.taxconnect.core.base.BaseActivity

import com.example.taxconnect.core.navigation.NavigationManager

class ExploreCAsActivity : BaseActivity<ActivityExploreCasBinding>(), CAAdapter.OnRequestAssistanceClickListener {
    override val bindingInflater: (LayoutInflater) -> ActivityExploreCasBinding = ActivityExploreCasBinding::inflate
    private val navigationManager = NavigationManager()
    private lateinit var repository: DataRepository
    private lateinit var adapter: CAAdapter
    private var fullCAList = ArrayList<UserModel>()
    private var filterMinPrice = 0.0
    private var filterMaxPrice = Double.MAX_VALUE
    private var currentSearchQuery = ""
    private var filterVerifiedOnly = false
    private var filterCityOnly = false
    private var favoritesOnly = false
    private var currentUserCity: String? = null
    private var favoriteIds = HashSet<String>()
    private var lastCaSnapshot: DocumentSnapshot? = null
    private var isLoadingCAs = false
    private var hasMoreCAs = true
    private val CA_PAGE_SIZE = 30

    private val searchHandler = Handler(Looper.getMainLooper())
    private val searchRunnable = Runnable { applySearch(currentSearchQuery) }
    private val prefs by lazy { getSharedPreferences("explore_prefs", MODE_PRIVATE) }

    companion object {
        private const val PREF_RECENT_SEARCHES = "recent_searches"
    }

    override fun initViews() {
        try {
            repository = DataRepository.getInstance()
        } catch (e: Exception) {
            timber.log.Timber.e(e, "Failed to initialize data repository")
            showToast("Failed to initialize data repository")
            finish()
            return
        }

        setupUI()
        setupRecyclerView()
        setupSorting()
        setupSearch()
        renderRecentChips(loadRecentSearches())
        showLoading(true)
        fetchCAList()
        loadFavorites()
        loadCurrentUserCity()

        val initialQuery = intent.getStringExtra("QUERY")
        if (!initialQuery.isNullOrEmpty()) {
            binding.etSearch.setText(initialQuery)
            binding.etSearch.setSelection(initialQuery.length)
            saveRecentSearch(initialQuery)
            renderRecentChips(loadRecentSearches())
        }
        favoritesOnly = intent.getBooleanExtra("FAVORITES_ONLY", false)
        binding.chipFavorites.isChecked = favoritesOnly
    }

    override fun observeViewModel() {
        // No ViewModel used in this activity currently
    }

    override fun setupListeners() {
        // Listeners are set up in setupUI
    }

    private fun setupUI() {
        binding.ivBack.setOnClickListener { finish() }
        binding.btnFilter.setOnClickListener { showPriceFilterDialog() }
        binding.chipVerified.setOnCheckedChangeListener { _, isChecked ->
            filterVerifiedOnly = isChecked
            filterList(currentSearchQuery)
        }
        binding.chipFavorites.setOnCheckedChangeListener { _, isChecked ->
            favoritesOnly = isChecked
            filterList(currentSearchQuery)
        }
        binding.chipMyCity.setOnCheckedChangeListener { buttonView, isChecked ->
            if (isChecked && currentUserCity.isNullOrBlank()) {
                buttonView.isChecked = false
                showToast(getString(R.string.add_city_profile_filter))
                return@setOnCheckedChangeListener
            }
            filterCityOnly = isChecked
            filterList(currentSearchQuery)
        }
    }

    private fun setupRecyclerView() {
        adapter = CAAdapter(ArrayList(), this, this, false)
        binding.rvCAList.layoutManager = LinearLayoutManager(this)
        binding.rvCAList.adapter = adapter
        binding.rvCAList.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                if (dy <= 0) return
                val layoutManager = recyclerView.layoutManager as? LinearLayoutManager ?: return
                val lastVisible = layoutManager.findLastVisibleItemPosition()
                val total = adapter.itemCount
                if (!isLoadingCAs && hasMoreCAs && lastVisible >= total - 5) {
                    loadMoreCAs(false)
                }
            }
        })
        adapter.setOnSaveClickListener(object : CAAdapter.OnSaveClickListener {
            override fun onSaveClick(ca: UserModel) {
                val uid = FirebaseAuth.getInstance().uid ?: return
                repository.toggleFavorite(uid, ca.uid ?: return, object : DataRepository.DataCallback<Boolean> {
                    override fun onSuccess(isFav: Boolean?) {
                        if (isFav == true) {
                            favoriteIds.add(ca.uid!!)
                        } else {
                            favoriteIds.remove(ca.uid!!)
                        }
                        adapter.setFavorites(favoriteIds)
                    }

                    override fun onError(error: String?) {}
                })
            }
        })
    }

    private fun showPriceFilterDialog() {
        val builder = AlertDialog.Builder(this)
        builder.setTitle(getString(R.string.filter_by_price))

        val layout = LinearLayout(this)
        layout.orientation = LinearLayout.VERTICAL
        layout.setPadding(32, 16, 32, 16)

        val inputMin = EditText(this)
        inputMin.inputType = InputType.TYPE_CLASS_NUMBER or InputType.TYPE_NUMBER_FLAG_DECIMAL
        inputMin.hint = "Min Price (₹)"
        if (filterMinPrice > 0) inputMin.setText(filterMinPrice.toString())
        layout.addView(inputMin)

        val inputMax = EditText(this)
        inputMax.inputType = InputType.TYPE_CLASS_NUMBER or InputType.TYPE_NUMBER_FLAG_DECIMAL
        inputMax.hint = "Max Price (₹)"
        if (filterMaxPrice < Double.MAX_VALUE) inputMax.setText(filterMaxPrice.toString())
        layout.addView(inputMax)

        builder.setView(layout)

        builder.setPositiveButton(getString(R.string.apply)) { _, _ ->
            val minStr = inputMin.text.toString()
            val maxStr = inputMax.text.toString()

            filterMinPrice = minStr.toDoubleOrNull() ?: 0.0
            filterMaxPrice = maxStr.toDoubleOrNull() ?: Double.MAX_VALUE

            filterList(currentSearchQuery)
        }

        builder.setNegativeButton(getString(R.string.clear)) { dialog, _ ->
            filterMinPrice = 0.0
            filterMaxPrice = Double.MAX_VALUE
            filterList(currentSearchQuery)
            dialog.cancel()
        }

        builder.setNeutralButton(getString(R.string.cancel)) { dialog, _ -> dialog.cancel() }

        builder.show()
    }

    private fun setupSorting() {
        val sortOptions = arrayOf("Rating (High to Low)", "Experience (High to Low)", "Charges (Low to High)")
        val sortAdapter = ArrayAdapter(this, R.layout.item_spinner, sortOptions)
        sortAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        binding.spinnerSort.adapter = sortAdapter

        binding.spinnerSort.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                sortList(position)
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }
    }

    private fun setupSearch() {
        binding.etSearch.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                currentSearchQuery = s.toString()
                debounceSearch()
            }
            override fun afterTextChanged(s: Editable?) {}
        })
    }

    private fun fetchCAList() {
        fullCAList.clear()
        lastCaSnapshot = null
        hasMoreCAs = true
        loadMoreCAs(true)
    }

    private fun loadMoreCAs(initialLoad: Boolean) {
        if (isLoadingCAs || !hasMoreCAs) return
        isLoadingCAs = true
        if (initialLoad) {
            showLoading(true)
        }

        repository.getCaPage(CA_PAGE_SIZE, lastCaSnapshot, object : DataRepository.DataCallback<DataRepository.PageResult<UserModel>> {
            override fun onSuccess(data: DataRepository.PageResult<UserModel>?) {
                if (initialLoad) showLoading(false)
                if (data == null) {
                    binding.layoutEmptyState.visibility = View.VISIBLE
                    binding.rvCAList.visibility = View.GONE
                    binding.tvRegisteredCount.text = getString(R.string.registered_cas_count)
                    isLoadingCAs = false
                    return
                }

                val newItems = data.items
                if (newItems.isEmpty()) {
                    hasMoreCAs = false
                } else {
                    fullCAList.addAll(newItems)
                    lastCaSnapshot = data.lastSnapshot
                }

                val fullText = getString(R.string.registered_cas_count_dynamic, fullCAList.size)
                val countText = fullCAList.size.toString()
                val spannable = SpannableString(fullText)
                val startIndex = fullText.indexOf(countText)
                if (startIndex != -1) {
                    spannable.setSpan(StyleSpan(Typeface.BOLD), startIndex, startIndex + countText.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                }
                binding.tvRegisteredCount.text = spannable

                if (fullCAList.isEmpty()) {
                    binding.layoutEmptyState.visibility = View.VISIBLE
                    binding.rvCAList.visibility = View.GONE
                } else {
                    binding.layoutEmptyState.visibility = View.GONE
                    binding.rvCAList.visibility = View.VISIBLE
                    filterList(currentSearchQuery)
                }
                isLoadingCAs = false
            }

            override fun onError(error: String?) {
                if (initialLoad) showLoading(false)
                isLoadingCAs = false
                handleError(error) { loadMoreCAs(initialLoad) }
            }
        })
    }

    private fun filterList(query: String) {
        currentSearchQuery = query
        val filtered = fullCAList.filter { ca ->
            val matchesQuery = query.isEmpty() ||
                    (ca.name?.contains(query, ignoreCase = true) == true) ||
                    (ca.specialization?.contains(query, ignoreCase = true) == true) ||
                    (ca.city?.contains(query, ignoreCase = true) == true)

            var matchesPrice = true
            ca.minCharges?.let { charges ->
                val price = parseCharges(charges)
                if (price != Double.MAX_VALUE) {
                    matchesPrice = price in filterMinPrice..filterMaxPrice
                }
            }

            val matchesVerified = !filterVerifiedOnly || ca.isVerified
            val matchesFavorites = !favoritesOnly || favoriteIds.contains(ca.uid)
            val matchesCity = !filterCityOnly || (currentUserCity != null && ca.city?.equals(currentUserCity, ignoreCase = true) == true)
            val matchesRole = ca.role == "CA"

            matchesQuery && matchesPrice && matchesVerified && matchesFavorites && matchesCity && matchesRole
        }

        if (filtered.isEmpty()) {
            binding.layoutEmptyState.visibility = View.VISIBLE
            binding.rvCAList.visibility = View.GONE
        } else {
            binding.layoutEmptyState.visibility = View.GONE
            binding.rvCAList.visibility = View.VISIBLE
            adapter.updateList(filtered)
        }
    }

    private fun sortList(position: Int) {
        if (fullCAList.isEmpty()) return

        when (position) {
            0 -> fullCAList.sortByDescending { it.rating }
            1 -> fullCAList.sortByDescending { parseExperience(it.experience) }
            2 -> fullCAList.sortBy { parseCharges(it.minCharges) }
        }
        filterList(currentSearchQuery)
    }

    private fun parseExperience(exp: String?): Double {
        if (exp == null) return 0.0
        return try {
            exp.replace("[^0-9.]".toRegex(), "").toDoubleOrNull() ?: 0.0
        } catch (e: Exception) {
            0.0
        }
    }

    private fun parseCharges(charge: String?): Double {
        if (charge == null) return Double.MAX_VALUE
        return try {
            charge.replace("[^0-9.]".toRegex(), "").toDoubleOrNull() ?: Double.MAX_VALUE
        } catch (e: Exception) {
            Double.MAX_VALUE
        }
    }

    private fun debounceSearch() {
        searchHandler.removeCallbacks(searchRunnable)
        searchHandler.postDelayed(searchRunnable, 300)
    }

    private fun applySearch(query: String) {
        filterList(query)
        if (query.trim().length >= 2) {
            saveRecentSearch(query.trim())
            renderRecentChips(loadRecentSearches())
        }
    }

    private fun loadRecentSearches(): List<String> {
        val csv = prefs.getString(PREF_RECENT_SEARCHES, "") ?: ""
        return if (csv.isEmpty()) emptyList() else csv.split("|").filter { it.isNotBlank() }
    }

    private fun saveRecentSearch(query: String) {
        val current = loadRecentSearches().toMutableList()
        current.remove(query)
        current.add(0, query)
        val ordered = if (current.size > 5) current.take(5) else current
        prefs.edit().putString(PREF_RECENT_SEARCHES, ordered.joinToString("|")).apply()
    }

    private fun clearRecentSearches() {
        prefs.edit().remove(PREF_RECENT_SEARCHES).apply()
        renderRecentChips(emptyList())
    }

    private fun renderRecentChips(items: List<String>) {
        binding.chipGroupRecent.removeAllViews()
        binding.hsRecentSearches.visibility = if (items.isEmpty()) View.GONE else View.VISIBLE
        for (label in items) {
            val chip = Chip(this)
            chip.text = label
            chip.isCheckable = false
            chip.isClickable = true
            chip.setOnClickListener {
                binding.etSearch.setText(label)
                binding.etSearch.setSelection(label.length)
                applySearch(label)
            }
            binding.chipGroupRecent.addView(chip)
        }
        if (items.isNotEmpty()) {
            val clearChip = Chip(this)
            clearChip.text = getString(R.string.clear)
            clearChip.isCheckable = false
            clearChip.isClickable = true
            clearChip.setOnClickListener { clearRecentSearches() }
            binding.chipGroupRecent.addView(clearChip)
        }
    }

    private fun showLoading(show: Boolean) {
        binding.shimmerLayout.visibility = if (show) View.VISIBLE else View.GONE
        binding.rvCAList.visibility = if (show) View.GONE else View.VISIBLE
        binding.tvEmptyState.visibility = View.GONE
    }

    private fun loadFavorites() {
        val uid = FirebaseAuth.getInstance().uid ?: return
        repository.getFavoriteCaIds(uid, object : DataRepository.DataCallback<List<String>> {
            override fun onSuccess(data: List<String>?) {
                if (data != null) {
                    favoriteIds.clear()
                    favoriteIds.addAll(data)
                    adapter.setFavorites(favoriteIds)
                    if (favoritesOnly) {
                        filterList(currentSearchQuery)
                    }
                }
            }

            override fun onError(error: String?) {}
        })
    }

    private fun loadCurrentUserCity() {
        val uid = FirebaseAuth.getInstance().uid ?: return
        repository.fetchUser(uid, object : DataRepository.DataCallback<UserModel> {
            override fun onSuccess(data: UserModel?) {
                data?.let {
                    currentUserCity = it.city
                    val hasCity = !currentUserCity.isNullOrBlank()
                    binding.chipMyCity.isEnabled = hasCity
                    binding.chipMyCity.visibility = if (hasCity) View.VISIBLE else View.GONE
                }
            }

            override fun onError(error: String?) {}
        })
    }

    override fun onRequestAssistanceClick(ca: UserModel) {
        navigationManager.navigateToCADetail(this, ca, true)
    }
}
